BZ-2062756 IPI Install Procedure Replacing the Control Plane Node (#43915)
openshift/openshift-docs: modules/ipi-install-replacing-a-bare-metal-control-plane-node.adoc

// This is included in the following assemblies:
//
// installing/installing_bare_metal_ipi/ipi-install-expanding-the-cluster.adoc

:_content-type: PROCEDURE
[id='replacing-a-bare-metal-control-plane-node_{context}']
= Replacing a bare metal control plane node

> Use the following procedure to replace an installer-provisioned {product-title} cluster.

Reviewer: Don't we mean 'to replace a bmh'?

Reviewer: Shouldn't the instructions include removing the old bmh, before adding a new one?

Reviewer 2: This is not about replacing a whole cluster, just a single control plane node.

Author: Updated as suggested.

[IMPORTANT]
====

> When the replacement BareMetalHost is created from a new physical machine or an incomplete prior YAML file configuration, the `externallyProvisioned:` field in the configuration file must be set to `false` to allow the cluster to provision the host. 

Reviewer: Thanks, Katie. 
Regarding externallyProvisioned: field, I think the issue in the customer case had been that it was set to true, but things should work smoothly even if it isn't set at all (meaning, it doesn't have to be set to 'false' explicitly). 
Either way, I don't think that sentence is very clear. Maybe 'previous, incomplete yaml file configuration' or something else that will make the meaning more clear?

> When the replacement BareMetalHost is created from a new physical machine or an incomplete prior YAML file configuration, the `externallyProvisioned:` field in the configuration file must be set to `false` to allow the cluster to provision the host. 

Reviewer 2: I agree it's not very clear. Adina is also correct that it's fine to leave the field unset, since `false` is the default.

A straightforward way to say this would be:

> If you reuse or copy-and-paste the BareMetalHost from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

That might be, uh, more straightforward than we're going for, idk.

> When the replacement BareMetalHost is created from a new physical machine or an incomplete prior YAML file configuration, the `externallyProvisioned:` field in the configuration file must be set to `false` to allow the cluster to provision the host. 

Author: Updated as suggested.

> Existing control plane BareMetalHosts may have the externallyProvisioned flag set to true if they were provisioned by the OpenShift Installer.

Reviewer 2: I imagine there's probably some `{product-title}`-type macros that are supposed to be used here, but I don't know the details.

====

.Procedure

. Create the baremetalhost and ipmi secret:

[source,terminal]
----
$ cat <<EOF | oc apply -f -
----

.Example output
[source,terminal]
----
apiVersion: v1
kind: Secret
metadata:
  name: master-3-bmc-secret
  namespace: openshift-machine-api
type: Opaque
data:
  username: _<username>_
  password: _<password>_
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: master-3
  namespace: openshift-machine-api
spec:
  externallyProvisioned: false
  online: true
  bootMACAddress: aa:aa:aa:aa:aa:03
  bmc:
	address: ipmi://172.22.0.1:6233
	credentialsName: master-3-bmc-secret
EOF
----

> . Once the inspection finishes, the bare metal host will be created and 'ready' to be provisioned:

Reviewer 2: The `ready` state was renamed to `available` in ~4.9 (don't quote me on the exact release)

worker-0	OK   	  provisioned          	  kni1-worker-0-kjpmp ipmi://10.19.143.29  	  dell-raid      	   true
----

. There are no machinesets for control plane nodes, so you'll need to create a machine object instead. The image urls and checksums can be copied from the other master objects:

[source,terminal]
----
$ cat <<EOF | oc apply -f -
----

.Example output
[source,terminal]
----
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  labels:
	machine.openshift.io/cluster-api-cluster: kni1
	machine.openshift.io/cluster-api-machine-role: master
	machine.openshift.io/cluster-api-machine-type: master
  name: kni1-master-3
  namespace: openshift-machine-api
spec:
  providerSpec:
	value:
  	hostSelector: {}
  	image:
    	checksum: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2.md5sum
    	url: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2
  	userData:
    	name: master-user-data
EOF
----

[NOTE]
====

> The baremetalhost, secret and machine objects can be created at once in a single command or yaml file if needed.

Reviewer 3: ```suggestion
You can define the `BareMetalHost`, `Secret`, and `MachineSet` objects in a single YAML file, for example, `sample.yaml` and create the objects using the following command:
[source,terminal]
----
$ oc create -f sample.yaml
----
```

====

. The provisioning process (via the baremetal-operator) will install RHCOS and prepare the host to be added to the cluster:

[source,terminal]
----
$ oc get bmh -A
----

.Example output
[source,terminal]
----
NAMESPACE           	  NAME        	 STATUS PROVISIONING STATUS  	  CONSUMER        BMC                  	  HARDWARE PROFILE   ONLINE   ERROR
openshift-machine-api   kni1-master-0  OK   	provisioned          	  kni1-worker-0   ipmi://10.19.136.22     dell-raid          true	 
openshift-machine-api   master-1   	   OK   	externally provisioned  kni1-master-1   ipmi://172.22.0.1:6231                     true	 
openshift-machine-api   master-2   	   OK   	externally provisioned  kni1-master-2   ipmi://172.22.0.1:6232                     true	 
openshift-machine-api   master-3   	   OK   	provisioning            kni1-master-3   ipmi://172.22.0.1:6233  unknown            true	 
openshift-machine-api   worker-0    	 OK   	provisioned          	  kni1-worker-1   ipmi://10.19.143.29  	  dell-raid          true

> ----

Reviewer 3: Adjust the output to render correctly in the preview. Presently it is a bit distorted.

[source,terminal]
----
$ oc get bmh -A
----

.Example output
[source,terminal]
----
NAMESPACE           	  NAME    	STATUS  PROVISIONING STATUS  	  CONSUMER          BMC                  	  HARDWARE PROFILE   ONLINE   ERROR
openshift-machine-api   master-1  OK   	  externally provisioned  kni1-master-1     ipmi://172.22.0.1:6231                     true	 
openshift-machine-api   master-2  OK    	externally provisioned  kni1-master-2     ipmi://172.22.0.1:6232                     true	 
openshift-machine-api   master-3  OK    	provisioned          	  kni1-master-3     ipmi://172.22.0.1:6233   unknown           true	 
openshift-machine-api   worker-0	OK    	provisioned          	  kni1-worker-0     ipmi://10.19.143.29  	   dell-raid         true
----

. Confirm that the node has been added to the cluster:

[source,terminal]
----
$ oc get nodes
----

.Example output
----
NAME                  STATUS      ROLES      AGE	  VERSION
kni1-master-1        	available	  master     141m   v1.18.2
kni1-master-2         available	  master     141m   v1.18.2
kni1-master-3         available	  master     4m2s   v1.18.2
kni1-worker-0         available	  worker     87m	  v1.18.2

> ----

Reviewer 3: Adjust the output to render correctly in the preview. Presently it is a bit distorted.

// This is included in the following assemblies:
//
// installing/installing_bare_metal_ipi/ipi-install-expanding-the-cluster.adoc

:_content-type: PROCEDURE
[id="replacing-a-bare-metal-control-plane-node_{context}"]
= Replacing a bare-metal control plane node

Use the following procedure to replace an installer-provisioned {product-title} cluster.

[IMPORTANT]

> ====

Reviewer 4: It is very important to note that the old bmh, machine and node objects need to be deleted before doing anything else.

Reviewer 4: ```
$ oc delete node kni1-vmaster-0
$ oc delete bmh -n openshift-machine-api vmaster-0
$ oc delete machine -n openshift-machine-api kni1-master-0
```

Author: Updated as suggested.

> When the replacement BareMetalHost is created from a new physical machine or an incomplete prior YAML file configuration, the `externallyProvisioned:` field in the configuration file must be set to `false` to allow the cluster to provision the host. 

Author: Updated as suggested.

> Existing control plane BareMetalHosts may have the externallyProvisioned flag set to true if they were provisioned by {product-title}.

Reviewer 2: I assume `{product-title}` expands to "OpenShift" here, but the key point is that control plane nodes are initially provisioned by **openshift-installer**, and thus are marked 'externallyProvisioned' when BMHs are retrospectively added for them in the cluster. This is as opposed to worker nodes which are provisioned through their BMHs. Arguably both control plane and worker nodes are provisioned by "OpenShift" though.

> Existing control plane BareMetalHosts may have the externallyProvisioned flag set to true if they were provisioned by {product-title}.

Author: Updated as suggested.

====

.Procedure

> 

Reviewer 4: Delete the bmh, machine and node

Author: Updated as suggested.

> . Create the `baremetalhost` and `ipmi` secret:

Author: Updated as suggested.

+
[source,terminal]
----
$ cat <<EOF | oc apply -f -

> ----

Reviewer 4: In 4.10 with virtual media, the bmh file look like:
```
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: ocp-master-2
  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78
    credentialsName: ocp-master-2-bmc-secret
  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown
  online: true
  ```

Author: Updated as suggested.

+
.Example input
[source,terminal]
----
apiVersion: v1
kind: Secret
metadata:
  name: master-3-bmc-secret
  namespace: openshift-machine-api
type: Opaque
data:
  username: _<username>_
  password: _<password>_
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: master-3
  namespace: openshift-machine-api
spec:
  externallyProvisioned: false
  online: true
  bootMACAddress: aa:aa:aa:aa:aa:03
  bmc:

> 	address: ipmi://172.22.0.1:6233

Author: Updated as suggested.

	credentialsName: master-3-bmc-secret

> EOF

Author: This makes sense to me, but I'll confirm with QE.

> EOF

Reviewer: I don't have a strong opinion either way. To me, either option conveys the method clearly enough.
However, I am wondering if we should find a way to say that the input is an example.

> EOF

Reviewer: To me, they both sound clear. So I think it's just a matter of convention.
However, maybe if we use the concatenated option, we should find a way to say that this is an example?

> EOF

Author: @awolffredhat combining the command and the concatenated file into a single step removes the "example input" heading. I generally prefer to call out examples, but I'm going to update this as Bob suggests, since that's in line with our other documentation.

> EOF

Author: Updated as suggested.

----
+

> . After the inspection is complete, the bare-metal host is created and `available` to be provisioned:

Author: Updated as suggested.

+
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]
----
NAME     STATUS STATE                  CONSUMER            BMC                    HARDWARE PROFILE ONLINE ERROR
master-1 OK     externally provisioned kni1-master-1       ipmi://172.22.0.1:6231                  true
master-2 OK     externally provisioned kni1-master-2       ipmi://172.22.0.1:6232                  true
master-3 OK     available                                  ipmi://172.22.0.1:6233 unknown          true
worker-0 OK     provisioned            kni1-worker-0-kjpmp ipmi://10.19.143.29    dell-raid        true
----
+

> . There are no `MachineSet` objects for control plane nodes, so you must create a `MachineSet` object instead. You can copy the image URL and checksum from the other BareMetalHost objects.

Author: Updated as suggested.

+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
----

> +

Reviewer 4: In 4.10 with virtual media:
```
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  annotations:
    metal3.io/BareMetalHost: openshift-machine-api/ocp-master-2
  labels:
    machine.openshift.io/cluster-api-cluster: ocp-chzt8
    machine.openshift.io/cluster-api-machine-role: master
    machine.openshift.io/cluster-api-machine-type: master
  name: ocp-chzt8-master-2
  namespace: openshift-machine-api
spec:
  metadata: {}
  providerSpec:
    value:
      apiVersion: baremetal.cluster.k8s.io/v1alpha1
      customDeploy:
        method: install_coreos
      hostSelector: {}
      image:
        checksum: ""
        url: ""
      kind: BareMetalMachineProviderSpec
      metadata:
        creationTimestamp: null
      userData:
        name: master-user-data-managed
```

Author: Updated as suggested.

.Example output
[source,terminal]
----
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  labels:
	machine.openshift.io/cluster-api-cluster: kni1
	machine.openshift.io/cluster-api-machine-role: master
	machine.openshift.io/cluster-api-machine-type: master
  name: kni1-master-3
  namespace: openshift-machine-api
spec:
  providerSpec:
	value:
  	hostSelector: {}
  	image:
    	checksum: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2.md5sum
    	url: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2
  	userData:
    	name: master-user-data

> EOF

Author: Updated as suggested.

> EOF

Author: Updated as suggested.

----
+

> [NOTE]

Author: Updated as suggested.

[source,terminal]
----
$ oc create -f sample.yaml
----
====
+
. The provisioning process (by using the baremetal-operator) installs RHCOS and prepares the host to be added to the cluster:
+
[source,terminal]
----
$ oc get bmh -A
----
+
.Example output
[source,terminal]
----
NAMESPACE               NAME           STATUS PROVISIONING STATUS     CONSUMER        BMC                     HARDWARE PROFILE   ONLINE   ERROR
openshift-machine-api   kni1-master-0  OK     provisioned             kni1-worker-0   ipmi://10.19.136.22     dell-raid          true
openshift-machine-api   master-1       OK     externally provisioned  kni1-master-1   ipmi://172.22.0.1:6231                     true
openshift-machine-api   master-2       OK     externally provisioned  kni1-master-2   ipmi://172.22.0.1:6232                     true
openshift-machine-api   master-3       OK     provisioning            kni1-master-3   ipmi://172.22.0.1:6233  unknown            true
openshift-machine-api   worker-0       OK     provisioned             kni1-worker-1   ipmi://10.19.143.29     dell-raid          true
----
+

> . The BareMetalHost is added to the cluster:

Reviewer 5: Because this is a step, it should be an action, for example:
```suggestion
. Verify that the `BareMetalHost` object is added to the cluster:
```

Author: Updated as suggested.

+
[source,terminal]
----
$ oc get bmh -A
----
+
.Example output
[source,terminal]
----
NAMESPACE             NAME     STATUS PROVISIONING STATUS     CONSUMER          BMC                    HARDWARE PROFILE ONLINE ERROR
openshift-machine-api master-1 OK     externally provisioned  kni1-master-1     ipmi://172.22.0.1:6231                  true 
openshift-machine-api master-2 OK     externally provisioned  kni1-master-2     ipmi://172.22.0.1:6232                  true
openshift-machine-api master-3 OK     provisioned             kni1-master-3     ipmi://172.22.0.1:6233 unknown          true
openshift-machine-api worker-0 OK     provisioned             kni1-worker-0     ipmi://10.19.143.29    dell-raid        true
----
+
. Confirm that the node has been added to the cluster:
+
[source,terminal]
----
$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME             STATUS      ROLES    AGE   VERSION
kni1-master-1    available	 master   141m  v1.18.2
kni1-master-2    available	 master   141m  v1.18.2
kni1-master-3    available	 master   4m2s  v1.18.2
kni1-worker-0    available	 worker   87m   v1.18.2

> ----

Reviewer 4: Then, fix etcd.

Get the list of etcd pods:

```
$ oc get po -n openshift-etcd -l app=etcd
NAME                               READY   STATUS             RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running            0          113m
etcd-kni1-vmaster-2                3/3     Running            0          111m
etcd-kni1-vmaster-3                2/3     CrashLoopBackOff   21         87m
```

Choose an etcd pod currently working and use it to restore the cluster (etcd-kni1-vmaster1 in this example)

Check the current etcd members, including the failed one:

```
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| 96614f1cfb922d6c | started | kni1-vmaster-0 | https://10.19.138.32:2380 | https://10.19.138.32:2379 |      false |
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

Remove the failed one and verify it is not longer there:
```
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member remove 96614f1cfb922d6c"
Member 96614f1cfb922d6c removed from cluster 33a400c7a5a9c48c

$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

After a while, the new node will be added to the cluster:

```
$ oc get po -n openshift-etcd -l app=etcd
NAME                               READY   STATUS              RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running             0          29s
etcd-kni1-vmaster-2                3/3     Running             0          116m
etcd-kni1-vmaster-3                3/3     Running             22         92m

$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
| fa5d82e7a7ab6ca8 | started | kni1-vmaster-3 | https://10.19.138.38:2380 | https://10.19.138.38:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

Author: @e-minguez do you have the commands and steps to choose a working etcd pod and using it to restore the cluster? It seems as though it's missing from your above write up, unless it's part of `oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"` and I'm just unaware? Thanks!

Reviewer 4: The pod in 'CrashLoopBack' is the one not working, so basically you can choose any other pod. I've edited the comment with a more consistent example (I mixed the instructions from a previous environment, apologize)

  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78
    credentialsName: kni1-master-0-bmc-secret
  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown
  online: true
EOF
----
+
After the inspection is complete, the `baremetalhost` object is created and available to be provisioned.

. View available `baremetalhost` objects:
+
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]
----
NAME           STATE                    CONSUMER                   ONLINE   ERROR   AGE
kni1-master-0  available                ocp-hkw9p-master-0         true             1h10m
kni1-master-1  externally provisioned   ocp-hkw9p-master-1         true             4h53m
kni1-master-2  externally provisioned   ocp-hkw9p-master-2         true             4h53m
kni1-worker-0  provisioned              ocp-hkw9p-worker-0-ktmmx   true             4h53m
kni1-worker-1  provisioned              ocp-hkw9p-worker-0-l2zmb   true             4h53m
----
+

> There are no `MachineSet` objects for control plane nodes, so you must create a `Machine` object instead. You can copy the image URL and checksum from the other `Machine` objects.

Reviewer 2: This is OK, but could be confusing given that in 4.10 we use a customDeploy method _instead_ of an image URL (including in the example below).

```suggestion
There are no `MachineSet` objects for control plane nodes, so you must create a `Machine` object instead. You can copy the `providerSpec` from another control plane `Machine` object.
```

Author: Updated as suggested.

  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78
    credentialsName: kni1-master-0-bmc-secret
  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown
  online: true
EOF
----
+
After the inspection is complete, the `baremetalhost` object is created and available to be provisioned.

. View available `baremetalhost` objects:
+
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]
----
NAME           STATE                    CONSUMER                   ONLINE   ERROR   AGE
kni1-master-0  available                ocp-hkw9p-master-0         true             1h10m
kni1-master-1  externally provisioned   ocp-hkw9p-master-1         true             4h53m
kni1-master-2  externally provisioned   ocp-hkw9p-master-2         true             4h53m
kni1-worker-0  provisioned              ocp-hkw9p-worker-0-ktmmx   true             4h53m
kni1-worker-1  provisioned              ocp-hkw9p-worker-0-l2zmb   true             4h53m
----
+

> There are no `MachineSet` objects for control plane nodes, so you must create a `MachineSet` object instead. You can copy the image URL and checksum from the other `BareMetalHost` objects.

Reviewer 4: ```suggestion
There are no `MachineSet` objects for control plane nodes, so you must create a `Machine` object instead. You can copy the image URL and checksum from the other `Machine` objects.
```

Author: Updated as suggested.

> . Create a `MachineSet` object:

Reviewer 4: ```suggestion
. Create a `Machine` object:
```

Author: Updated as suggested.

+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  annotations:
    metal3.io/BareMetalHost: openshift-machine-api/kni1-master-0
  labels:
    machine.openshift.io/cluster-api-cluster: kni1
    machine.openshift.io/cluster-api-machine-role: master
    machine.openshift.io/cluster-api-machine-type: master
  name: kni1-master-0
  namespace: openshift-machine-api
spec:
  metadata: {}
  providerSpec:
    value:
      apiVersion: baremetal.cluster.k8s.io/v1alpha1
      customDeploy:
        method: install_coreos
      hostSelector: {}
      image:
        checksum: ""
        url: ""
      kind: BareMetalMachineProviderSpec
      metadata:
        creationTimestamp: null
      userData:
        name: master-user-data-managed
EOF
----
+

> . To define and create the `BareMetalHost`, `Secret`, and `MachineSet` objects in a single step, create a YAML file (`example.yaml`) with their definitions and run the following command:

Reviewer 4: ```suggestion
. To define and create the `BareMetalHost`, `Secret`, and `Machine` objects in a single step, create a YAML file (`example.yaml`) with their definitions and run the following command:
```

Author: Updated as suggested.

+
[source,terminal]
----
$ oc create -f example.yaml
----
+
The provisioning process uses the baremetal-operator to install RHCOS and prepare the host to be added to the cluster. 
+

> . To view the `baremetalhost` objects, run the following command:

Reviewer 4: ```suggestion
. To view the `BareMetalHost` objects, run the following command:
```
For consistency as it is using `BareMetalHost` in all the other sentences

Author: Updated as suggested.

+
[source,terminal]
----
$ oc get bmh -A
----
+
.Example output
[source,terminal]
----
NAME           STATE                    CONSUMER                   ONLINE   ERROR   AGE
kni1-master-0  provisioned              ocp-hkw9p-master-0         true             2h53m
kni1-master-1  externally provisioned   ocp-hkw9p-master-1         true             5h53m
kni1-master-2  externally provisioned   ocp-hkw9p-master-2         true             5h53m
kni1-worker-0  provisioned              ocp-hkw9p-worker-0-ktmmx   true             5h53m
kni1-worker-1  provisioned              ocp-hkw9p-worker-0-l2zmb   true             5h53m
----
+
. After the RHCOS installation, verify that the `BareMetalHost` is added to the cluster:
+

> [source,terminal]

Reviewer 4: This command is duplicated

Author: Updated as suggested.

----
$ oc get bmh -A
----
+
. Confirm that the node has been added to the cluster:
+
[source,terminal]
----
$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME             STATUS      ROLES    AGE   VERSION
kni1-master-0    available	 master   4m2s  v1.18.2
kni1-master-1    available	 master   141m  v1.18.2
kni1-master-2    available	 master   141m  v1.18.2
kni1-worker-0    available	 worker   87m   v1.18.2
----
+
[NOTE]
====

> After replacement of the new control plane node, the etcd pod is in `crashloopback` status. See "Replacing an unhealthy etcd member" for more information.

Reviewer 4: ```suggestion
After replacement of the new control plane node, the etcd pod running in the new node is in `crashloopback` status. See "Replacing an unhealthy etcd member" for more information.
```

Author: Updated as suggested.

// This is included in the following assemblies:
//
// installing/installing_bare_metal_ipi/ipi-install-expanding-the-cluster.adoc

:_content-type: PROCEDURE
[id="replacing-a-bare-metal-control-plane-node_{context}"]
= Replacing a bare-metal control plane node

Use the following procedure to replace an installer-provisioned {product-title} control plane node.

[IMPORTANT]
====

> If you reuse the `BareMetalHost` objects from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

Reviewer 4: ```suggestion
If you reuse the `BareMetalHost` object definition from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.
```

Author: Updated as suggested.

> Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by {product-title}.

Reviewer 2: Not sure if you missed [this comment](https://github.com/openshift/openshift-docs/pull/43915#discussion_r847697613), or if I'm just wrong about what `{product-title}` expands to. If the former then this should be something like:

```suggestion
Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by the {product-title} installer.
```

Author: I believe you're correct. That said, I'll request another peer review, as this module has changed quite a bit since they last saw it.

Author: Updated to "the {product-title} installation program", per rh style guidelines.

====

.Procedure

> . Remove the old `baremetalhost`, machine and node objects:

Reviewer 4: ```suggestion
. Remove the old `baremetalhost`, `machine` and `node` objects:
```

Author: Updated as suggested.

+
[source,terminal]
----
$ oc delete node kni1-vmaster-0
$ oc delete bmh -n openshift-machine-api vmaster-0
$ oc delete machine -n openshift-machine-api kni1-master-0
----
+
. Create the new `baremetalhost` object and the secret to store the BMC credentials:
+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
apiVersion: v1
kind: Secret
metadata:

>   name: ocp-master-2-bmc-secret

Author: Updated as suggested.

  namespace: openshift-machine-api
data:
  password: <username>
  username: <password>
type: Opaque
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:

>   name: ocp-master-2

Author: Updated as suggested.

  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78

>     credentialsName: ocp-master-2-bmc-secret

Author: Updated as suggested.

  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown
  online: true
EOF
----
+
After the inspection is complete, the `baremetalhost` object is created and available to be provisioned.

. View available `baremetalhost` objects:
+
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]

> ----

Reviewer 4: The BMC urls don't match... the ipmi ones were from an example I did back in the day, now I've switched to redfish... maybe you can use:

```
NAME           STATE                    CONSUMER                   ONLINE   ERROR   AGE
kni1-master-0  available                ocp-hkw9p-master-0         true             1h10m
kni1-master-1  externally provisioned   ocp-hkw9p-master-1         true             4h53m
kni1-master-2  externally provisioned   ocp-hkw9p-master-2         true             4h53m
kni1-worker-0  provisioned              ocp-hkw9p-worker-0-ktmmx   true             4h53m
kni1-worker-1  provisioned              ocp-hkw9p-worker-0-l2zmb   true             4h53m

Author: Updated as suggested.

NAME     STATUS STATE                  CONSUMER            BMC                    HARDWARE PROFILE ONLINE ERROR
master-1 OK     externally provisioned kni1-master-1       ipmi://172.22.0.1:6231                  true
master-2 OK     externally provisioned kni1-master-2       ipmi://172.22.0.1:6232                  true

> master-3 OK     available                                  ipmi://172.22.0.1:6233 unknown          true

Author: Updated as suggested.

worker-0 OK     provisioned            kni1-worker-0-kjpmp ipmi://10.19.143.29    dell-raid        true
----
+
There are no `MachineSet` objects for control plane nodes, so you must create a `MachineSet` object instead. You can copy the image URL and checksum from the other `BareMetalHost` objects.

. Create a `MachineSet` object:
+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  annotations:

>     metal3.io/BareMetalHost: openshift-machine-api/ocp-master-2

Author: Updated as suggested.

  labels:
    machine.openshift.io/cluster-api-cluster: kni1
    machine.openshift.io/cluster-api-machine-role: master
    machine.openshift.io/cluster-api-machine-type: master

>   name: kni1-master-2

Reviewer 4: ```suggestion
  name: kni1-master-0
```

Author: Updated as suggested.

  namespace: openshift-machine-api
spec:
  metadata: {}
  providerSpec:
    value:
      apiVersion: baremetal.cluster.k8s.io/v1alpha1
      customDeploy:
        method: install_coreos
      hostSelector: {}
      image:
        checksum: ""
        url: ""
      kind: BareMetalMachineProviderSpec
      metadata:
        creationTimestamp: null
      userData:
        name: master-user-data-managed
EOF
----
+

> . To define and create the `BareMetalHost`, `Secret`, and `MachineSet` objects in a single YAML file, run the following command:

Author: Updated as suggested.

+
[source,terminal]
----
$ oc create -f example.yaml
----
+
The provisioning process uses the baremetal-operator to install RHCOS and prepare the host to be added to the cluster. 
+
. To view the `baremetalhost` objects, run the following command:
+
[source,terminal]
----
$ oc get bmh -A
----
+
.Example output
[source,terminal]

> ----

Author: Updated as suggested.

NAMESPACE               NAME           STATUS PROVISIONING STATUS     CONSUMER        BMC                     HARDWARE PROFILE   ONLINE   ERROR
openshift-machine-api   kni1-master-0  OK     provisioned             kni1-worker-0   ipmi://10.19.136.22     dell-raid          true
openshift-machine-api   master-1       OK     externally provisioned  kni1-master-1   ipmi://172.22.0.1:6231                     true
openshift-machine-api   master-2       OK     externally provisioned  kni1-master-2   ipmi://172.22.0.1:6232                     true
openshift-machine-api   master-3       OK     provisioning            kni1-master-3   ipmi://172.22.0.1:6233  unknown            true
openshift-machine-api   worker-0       OK     provisioned             kni1-worker-1   ipmi://10.19.143.29     dell-raid          true
----
+

> . Verify that the `BareMetalHost` is added to the cluster:

Reviewer 4: ```suggestion
. After the RHCOS installation, the node will be added to the cluster.
```

Author: Updated, but reworded to present tense.

> +

Reviewer 4: There is no need to show this output, it would be the same than the previous one.

Author: Removed output as suggested.

[source,terminal]
----
$ oc get bmh -A
----
+
.Example output
[source,terminal]
----
NAMESPACE             NAME     STATUS PROVISIONING STATUS     CONSUMER          BMC                    HARDWARE PROFILE ONLINE ERROR
openshift-machine-api master-1 OK     externally provisioned  kni1-master-1     ipmi://172.22.0.1:6231                  true 
openshift-machine-api master-2 OK     externally provisioned  kni1-master-2     ipmi://172.22.0.1:6232                  true
openshift-machine-api master-3 OK     provisioned             kni1-master-3     ipmi://172.22.0.1:6233 unknown          true
openshift-machine-api worker-0 OK     provisioned             kni1-worker-0     ipmi://10.19.143.29    dell-raid        true
----
+
. Confirm that the node has been added to the cluster:
+
[source,terminal]
----
$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME             STATUS      ROLES    AGE   VERSION
kni1-master-1    available	 master   141m  v1.18.2
kni1-master-2    available	 master   141m  v1.18.2

> kni1-master-3    available	 master   4m2s  v1.18.2

Reviewer 4: ```suggestion
kni1-master-0    available	 master   4m2s  v1.18.2
```

Reviewer 4: And move it to the top of this output because it is sorted

Author: Updated as suggested.

// This is included in the following assemblies:
//
// installing/installing_bare_metal_ipi/ipi-install-expanding-the-cluster.adoc

:_content-type: PROCEDURE
[id="replacing-a-bare-metal-control-plane-node_{context}"]
= Replacing a bare-metal control plane node

Use the following procedure to replace an installer-provisioned {product-title} control plane node.

[IMPORTANT]
====
If you reuse the `BareMetalHost` objects from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by {product-title}.
====

.Procedure

. Remove the old `baremetalhost`, machine and node objects:
+
[source,terminal]
----
$ oc delete node kni1-vmaster-0
$ oc delete bmh -n openshift-machine-api vmaster-0
$ oc delete machine -n openshift-machine-api kni1-master-0
----
+
. Create the new `baremetalhost` object and the secret to store the BMC credentials:
+
[source,terminal]
----

> $ cat <<EOF | oc apply -f -

Reviewer 4: This is not ok. You need to create both the secret to store the BMC credentials AND the `baremetalhost` object:

```
$ cat <<EOF | oc apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: ocp-master-2-bmc-secret
  namespace: openshift-machine-api
data:
  password: Y2Fsdmlu
  username: cm9vdA==
type: Opaque
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: ocp-master-2
  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78
    credentialsName: ocp-master-2-bmc-secret
  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown
  online: true
EOF
```

----
+
. Verify the `etcd` pod member list:
+
[source,terminal]
----
$ oc get po -n openshift-etcd -l app=etcd
----
+
.Example output
[source,terminal]
----
NAME                               READY   STATUS             RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running            0          113m
etcd-kni1-vmaster-2                3/3     Running            0          111m
etcd-kni1-vmaster-3                2/3     CrashLoopBackOff   21         87m
----
+
. Restore the cluster using one of the running `etcd` pods.
+
. Display the current `etcd` pod member details:
+
[source,terminal]
----
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
----
+
.Example output
[source,terminal]
----
ID                STATUS    NAME            PEER ADDRS                 CLIENT ADDRS               IS LEARNER
ccf1a165dac53790  started   kni1-vmaster-1  https://10.19.138.33:2380  https://10.19.138.33:2379  false 
e58335b2a6977818  started   kni1-vmaster-2  https://10.19.138.37:2380  https://10.19.138.37:2379  false
96614f1cfb922d6c  started   kni1-vmaster-3  https://10.19.138.32:2380  https://10.19.138.32:2379  false
----
+

> . Remove the failed `etcd` pod:

Reviewer 4: ```suggestion
. Remove the failed `etcd` member of the already replaced host:
```

+
[source,terminal]
----
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member remove 96614f1cfb922d6c"
----
+
.Example output
[source,terminal]
----
Member 96614f1cfb922d6c removed from cluster 33a400c7a5a9c48c
----
+

> . Verify that the failed `etcd` pod no longer exists:

Reviewer 4: ```suggestion
. Verify that the failed `etcd` member no longer exists:
```

+
[source,terminal]
----
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
----
+
.Example output
[source,terminal]
ID                STATUS  NAME            PEER ADDRS                 CLIENT ADDRS IS            LEARNER
ccf1a165dac53790  started kni1-vmaster-1  https://10.19.138.33:2380  https://10.19.138.33:2379  false
e58335b2a6977818  started kni1-vmaster-2  https://10.19.138.37:2380  https://10.19.138.37:2379  false
----
+

> . Verify the newly updated `etcd` pod member list:

Reviewer 4: ```suggestion
. After a few moments, the new member will be added automatically to the cluster. Verify the newly updated `etcd` pods list:
```

+
[source,terminal]
----
$ oc get po -n openshift-etcd -l app=etcd
----
+
.Example output
[source,terminal]
----
NAME                               READY   STATUS             RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running            0          29s
etcd-kni1-vmaster-2                3/3     Running            0          116m
etcd-kni1-vmaster-3                3/3     Running            22         92m
----
+

> . Display the current `etcd` pod member details:

Reviewer 4: ```suggestion
. Display the current `etcd` member details:
```
